// salesforcedx-templates/lib/utils/templateCommand.js
"use strict";
/*
 * Copyright (c) 2019, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const command_1 = require("@salesforce/command");
const core_1 = require("@salesforce/core");
const path = require("path");
// @ts-ignore
const yeoman = require("yeoman-environment");
const adapter_1 = require("./adapter");
const messageUtil_1 = require("./messageUtil");
class TemplateCommand extends command_1.SfdxCommand {
    static buildJson(adapter, targetDir) {
        const cleanOutput = adapter.log.getCleanOutput();
        const rawOutput = `target dir = ${targetDir}\n${adapter.log.getOutput()}`;
        const output = {
            outputDir: targetDir,
            created: cleanOutput,
            rawOutput
        };
        return output;
    }
    static getDefaultApiVersion() {
        const packageJsonPath = path.join('..', '..', 'package.json');
        const versionTrimmed = require(packageJsonPath).version.trim();
        return `${versionTrimmed.split('.')[0]}.0`;
    }
    static getApiVersion() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            try {
                const aggregator = yield core_1.ConfigAggregator.create();
                const apiVersionFromConfig = aggregator.getPropertyValue(TemplateCommand.API_VERSION);
                return apiVersionFromConfig || TemplateCommand.getDefaultApiVersion();
            }
            catch (err) {
                return TemplateCommand.getDefaultApiVersion();
            }
        });
    }
    runGenerator(generator) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            // Can't specify a default value the normal way for apiversion, so set it here
            if (!this.flags.apiversion) {
                this.flags.apiversion = yield TemplateCommand.getApiVersion();
            }
            const adapter = new adapter_1.ForceGeneratorAdapter();
            // @ts-ignore
            const env = yeoman.createEnv(undefined, undefined, adapter);
            env.registerStub(generator, 'generator');
            const result = yield env.run('generator', this.flags);
            const targetDir = path.resolve(this.flags.outputdir);
            if (this.flags.json) {
                return TemplateCommand.buildJson(adapter, targetDir);
            }
            else {
                this.log(messageUtil_1.MessageUtil.get('TargetDirOutput', [targetDir]));
                this.log(adapter.log.getOutput());
                return result;
            }
        });
    }
}
exports.TemplateCommand = TemplateCommand;
TemplateCommand.API_VERSION = 'apiVersion';
